content/post/2020-02-07-processx-blocked-sigchld-fix/index.Rmd

```diff
@@ -0,0 +1,442 @@
+---
+slug: processx-blocked-sigchld
+title: "Debugging: Signals and Subprocesses"
+authors:
+  - Gábor Csárdi
+date: "2020-02-07"
+tags:
+- package development
+- debugging
+output: 
+  html_document:
+    keep_md: true
+---
+
+```{r include = FALSE}
+options(width = 60)
+```
+
+This is a short story about a non-trivial bug in the processx package,
+and how I fixed it. It is a good showcase of the some debugging tools.
+
+## The bug
+
+[processx](https://processx.r-lib.org) is an R package to start and manage
+external processes. It is used by the [callr](https://call.r-lib.org)
+package to run code in another R session. The
+[original bug](https://github.com/r-lib/processx/issues/240) report has a
+nice, clean [reproducible example](https://reprex.tidyverse.org/):
+
+> https://github.com/r-lib/processx/pull/237 seems to solve
+> https://github.com/r-lib/processx/issues/236 for `processx`,
+> but not for `callr`.
+> 
+> ```r
+> fun <- function() {
+>   parallel::mclapply(1:2, function(x) x)
+> }
+> env <- c(
+>   callr::rcmd_safe_env(),
+>   PROCESSX_NOTIFY_OLD_SIGCHLD = "true"
+> )
+> tmp <- callr::r(fun, env = env, show = TRUE)
+> #> Error while shutting down parallel: unable to terminate some child processes
+```
+
+## A red herring
+
+The references to [PR #237](https://github.com/r-lib/processx/pull/237) and
+[issue #236](https://github.com/r-lib/processx/issues/236) are a red
+herring, and a tricky one! The error message is exactly the same there,
+but the reasons are very different. That issue is about an interference
+between processx and the parallel package. This one must be something else,
+because callr does not load processx (or any other R package) in the R
+subprocess it creates:
+```{r}
+callr::r(function() loadedNamespaces())
+```
+
+## About `SIGCHLD` signals
+
+To understand what is going on here, we need to know a bit about signals
+and subprocesses. A signal is an asynchronous notification, sent to a
+process by a(nother) process or the operating system. A signal means
+that an important event has happened, and the process's normal execution
+is interrupted. For some signal types, the process gets a chance to handle
+the signal gracefully and then continue execution. The `SIGCHLD` signal is
```

Comment:
State that it is sig like signal and chld like child if that's the case? And are all signal names allcaps?